### PR TITLE
Add Residual Bottleneck Block

### PR DESCRIPTION
**Pull request by niniack**

### CODE REVIEW

**Comment by niniack** [comment]
Notes:
- This PR implements a 'ResidualBlockType' class to avoid using strings when specifiying what block type we want to use. The hope is to avoid silly typos.
- There are "variations" to bottlenecks ([see here](https://dhruvs.space/posts/understanding-resnets/#bottleneck-residual-block)). I am using variation 1, where the earlier conv2d output channels is divided by a factor, rather than the flavor where the later conv2d output channel is multiplied by a factor. This allows us to bottleneck without worrying about messing with downstream filter sizes.
- I am not super keen on how the logic looks like in `create_block`, its heading towards a bit of an "if/else" mess. Maybe we reconsider how that looks if its going to get any more complicated.

File: src/model.py [hunk 1]
```diff
@@ -50,6 +57,64 @@ def forward(self, inputs: torch.Tensor) -> torch.Tensor:
...
+        '''
+        super().__init__()
+        self.num_channels = num_channels
+        self.use_stem = use_stem
+        self.strides = strides
+        self.factor = factor
+
+        # First convolutional layer with normalization
+        self.conv1 = nn.LazyConv2d(
+            num_channels//factor, kernel_size=1, padding=0)
```
Comment (ab7289) [inline src/model.py]: I think the decision to downsample by a factor is a great idea. That way we can specify different downsample factors if we want.

File: src/model.py [hunk 2]
```diff
@@ -50,6 +57,64 @@ def forward(self, inputs: torch.Tensor) -> torch.Tensor:
         return self.out(x)
 
 
+class ResidualBottleNeck(nn.Module):
```
Comment (ab7289) [inline src/model.py]: once @os0000os approves my PR we should include dropout in the block as well.

Comment (os0000os) [inline src/model.py]: Could we rename this class as `BottleneckResidualBlock`? I know the textbook uses this name, but it just sounds like it's referring to the bottleneck layers *within* a block.

**Review by ab7289 — APPROVED** [review]
LGTM!
